js/gacha.js

```diff
@@ -1,10 +1,99 @@
 import { renderCards } from "./cards.js";
+import { cards_by_rarity } from "./main.js";
 
 export const GACHA_BUTTON = document.getElementById("gacha-button");
-export const CARDS_PER_PULL = 10;
 
-//Pulls n cards from the passed "cards" array using the Fisher-Yates Shuffle.
-function pullCards(cards, n) {
+//Card slots on each pull, representing the percent chance of getting a card of that rarity in each slot.
+//The sum of all rarities on a slot should be 100 or higher for proper function.
+const slots = [
+  {
+    Common: 100,
+  },
+  {
+    Common: 100,
+  },
+  {
+    Common: 80,
+    Uncommon: 20,
+  },
+  {
+    Common: 80,
+    Uncommon: 20,
+  },
+  {
+    Uncommon: 100,
+  },
+  {
+    Uncommon: 100,
+  },
+  {
+    Uncommon: 26,
+    Rare: 26,
+    HoloRare: 21,
+    UltraRare: 17,
+    SecretRare: 10,
+  },
+  {
+    Uncommon: 26,
+    Rare: 26,
+    HoloRare: 21,
+    UltraRare: 17,
+    SecretRare: 10,
+  },
+  {
+    Rare: 40,
+    HoloRare: 20,
+    UltraRare: 20,
+    SecretRare: 20,
+  },
+  {
+    Rare: 40,
+    HoloRare: 20,
+    UltraRare: 20,
+    SecretRare: 20,
+  },
+];
+
+//Slots that replace one or more normal slots on special conditions.
+//This can be used to help collect all cards.
+const specialSlots = {
+  69: {
+    HoloRare: 100 / 3,
+    UltraRare: 100 / 3,
+    SecretRare: 100 / 3,
+  },
+};
+
+//Pulls a set of cards guided by an array of "slots" passed to the function.
+function pullCards(slots) {
+  let cards = [];
+  for (let slot of slots) {
+    let dice = Math.floor(Math.random() * 100) + 1;
+    if (dice in specialSlots) {
+      slot = specialSlots[dice];
+    }
+    for (let rarity in slot) {
+      if (dice <= slot[rarity]) {
+        cards.push(getRandomCards(cards_by_rarity[rarity], 1)[0]);
+        break;
+      } else {
+        dice -= slot[rarity];
+      }
```

Comment:
so the higher the dice value, the rarer the card?
Might be helpful to comment that to help readers understand the algorithm more easily

Reply:
Yes, basically. It's a little more complex, but the comment should explain it better now, it's a pretty long comment lol